resource-managers/kubernetes/token-refresh-server/src/main/scala/org/apache/spark/security/kubernetes/TokenRefreshService.scala

```diff
@@ -0,0 +1,361 @@
+    val maybeNewToken = if (token.getKind.equals(DelegationTokenIdentifier.HDFS_DELEGATION_KIND)) {
+      // The token can casted to AbstractDelegationTokenIdentifier below only if the token kind
+      // is HDFS_DELEGATION_KIND, according to the YARN resource manager code. See if this can be
+      // generalized beyond HDFS tokens.
+      val identifier = token.decodeIdentifier().asInstanceOf[AbstractDelegationTokenIdentifier]
+      val maxDate = identifier.getMaxDate
+      if (maxDate - expireTime < RENEW_TASK_REMAINING_TIME_BEFORE_NEW_TOKEN_MILLIS ||
+        maxDate <= nowMills) {
+        logDebug(s"Obtaining a new token with maxData $maxDate," +
+          s" expireTime $expireTime, now $nowMills")
+        val newToken = obtainNewToken(token, identifier)
+        logInfo(s"Obtained token $newToken")
+        newToken
+      } else {
+        token
+      }
+    } else {
+      token
+    }
+    maybeNewToken
+  }
+
+  private def maybeGetNewExpireTime(token: Token[_ <: TokenIdentifier], expireTime: Long,
+                                    deadline: Long,
+                                    nowMillis: Long) = {
+    if (expireTime <= deadline || expireTime <= nowMillis) {
+      try {
+        logDebug(s"Renewing token $token with current expire time $expireTime," +
+          s" deadline $deadline, now $nowMillis")
+        val newExpireTime = token.renew(hadoopConf)
+        logDebug(s"Renewed token $token. Next expire time $newExpireTime")
+        newExpireTime
+      } catch {
+        case t: Throwable =>
+          logWarning(t.getMessage, t)
+          hasError = true
+
+          getRetryTime
+      }
+    } else {
+      expireTime
+    }
+  }
+
+  private def obtainNewToken(token: Token[_ <: TokenIdentifier],
+                             identifier: AbstractDelegationTokenIdentifier) = {
+    val owner = identifier.getOwner
+    val realUser = identifier.getRealUser
+    val user = if (realUser == null || realUser.toString.isEmpty || realUser.equals(owner)) {
+      owner.toString
+    } else {
+      realUser.toString
+    }
+    val credentials = new Credentials()
+    val ugi = UserGroupInformation.createProxyUser(user, UserGroupInformation.getLoginUser)
+    val newToken = ugi.doAs(new PrivilegedExceptionAction[Token[_ <: TokenIdentifier]] {
+
+      override def run() : Token[_ <: TokenIdentifier] = {
+        val fs = FileSystem.get(hadoopConf)
+        val tokens = fs.addDelegationTokens(UserGroupInformation.getLoginUser.getUserName,
+          credentials)
+        tokens(0)
+      }
+    })
+    newToken
+  }
+
+  private def writeTokensToSecret(tokenToExpire: Map[Token[_ <: TokenIdentifier], Long],
+                                  nowMillis: Long) : Unit = {
+    val durationUntilExpire = tokenToExpire.values.min - nowMillis
+    val key = s"$SECRET_DATA_ITEM_KEY_PREFIX_HADOOP_TOKENS$nowMillis-$durationUntilExpire"
+    val credentials = new Credentials()
```

Comment:
Should this be blank? or should this take in the users HadoopConf from his `hdfs-site.xml`

Reply:
I don't think `Credentials` constructor take in a hadoop config object.